FGEGraphics/UISystem/UIInputLabel.cs

```diff
@@ -0,0 +1,469 @@
+
+    /// <summary>Adds text given two selection indices.</summary>
+    /// <param name="text">The text to add.</param>
+    /// <param name="indexLeft">The left index position.</param>
+    /// <param name="indexRight">The right index position.</param>
+    public void AddText(string text, int indexLeft, int indexRight)
+    {
+        string result = TextContent[..indexLeft] + text + TextContent[indexRight..];
+        Internal.CursorRight = Internal.CursorLeft += text.Length;
+        EditText(EditType.Add, text, result);
+    }
+
+    /// <summary>Deletes text between two indices.</summary>
+    /// <param name="indexLeft">The left index position.</param>
+    /// <param name="indexRight">The right index position.</param>
+    public void DeleteText(int indexLeft, int indexRight)
+    {
+        string diff = TextContent[indexLeft..indexRight];
+        string result = TextContent[..indexLeft] + TextContent[indexRight..];
+        EditText(EditType.Delete, diff, result, () => Internal.SetPosition(indexLeft));
+    }
+
+    /// <summary>Submits the current text content.</summary>
+    public void SubmitText() => EditText(EditType.Submit, string.Empty, TextContent);
+
+    /// <summary>Deletes text based on the <see cref="KeyHandlerState.InitBS"/> value.</summary>
+    /// <param name="keys">The current keyboard state.</param>
+    public void TickBackspaces(KeyHandlerState keys)
+    {
+        if (keys.InitBS == 0 || TextContent.Length == 0 || Internal.IndexRight == 0)
+        {
+            return;
+        }
+        if (Internal.IsSelection)
+        {
+            DeleteText(Internal.IndexLeft, Internal.IndexRight);
+            keys.InitBS--;
+        }
+        if (keys.InitBS > 0)
+        {
+            int index = Math.Max(Internal.IndexLeft - keys.InitBS, 0);
+            DeleteText(index, Internal.IndexRight);
+        }
+    }
+
+    /// <summary>Adds text based on the <see cref="KeyHandlerState.KeyboardString"/> value.</summary>
+    /// <param name="keys">The current keyboard state.</param>
+    public void TickContent(KeyHandlerState keys)
+    {
+        if (keys.KeyboardString.Length > 0)
+        {
+            AddText(keys.KeyboardString, Internal.IndexLeft, Internal.IndexRight);
+        }
+    }
+
+    // TODO: Handle ctrl left/right, handle up/down arrows
+    /// <summary>Modifies the current selection based on the <see cref="KeyHandlerState.LeftRights"/> value.</summary>
+    /// <param name="keys">The current keyboard state.</param>
+    /// <param name="shiftDown">Whether the shift key is being held.</param>
+    public void TickArrowKeys(KeyHandlerState keys, bool shiftDown)
+    {
+        if (keys.LeftRights == 0)
+        {
+            return;
+        }
+        bool wasSelection = Internal.IsSelection && !shiftDown;
+        Internal.CursorRight = keys.LeftRights < 0
+            ? (wasSelection ? Internal.IndexLeft : Math.Max(Internal.CursorRight + keys.LeftRights, 0))
```

Comment:
whaaaat is this? just move `CursorEnd += LeftRights` (or `-` whichever of the two matches the direction)